proposals/0370-modifiers.rst

```diff
@@ -44,136 +45,246 @@ a different syntax. Specifically, the term-level operators ``inline`` and
-     gadt_constrs ::= {modifier} con_list '::' sigtype
+     gadt_constrs ::= modifiers con_list '::' sigtype
 
    Modifiers in ``gadt_constrs`` apply to each constructor in ``con_list``.
 
-9. Reserve the use of ``%`` in a prefix occurrence to be used only for modifiers;
-   though this proposal does not do so, we can imagine extending the modifier syntax
-   to apply to further syntactic situations (e.g. term-level operators, declarations,
-   import lists, etc.).
+7. Reserve the use of ``%`` in a prefix occurrence to be used only for
+   modifiers; though this proposal does not do so, we can imagine extending the
+   modifier syntax to apply to further syntactic situations (e.g. term-level
+   operators, declarations, import lists, etc.).
+
+8. Modifiers are parsed, renamed, and type-checked as *types*.
+
+9. With ``-XModifiers``, the type of a modifier is determined only by synthesis,
+   not by checking. That is, in the bidirectional type-checking scheme used by
+   GHC, we find the type of the modifier by running the synthesis judgment.
+   Effectively, this means that if we consider a modifier to be some head
+   (constructor or variable) applied to a sequence of arguments (possibly none),
+   the head must have a known type: constructors always have a known type, and
+   variables have a known type if declared with a type signature. Alternatively,
+   the modifier may have a top-level type signature.
+
+10. Future modifiers will be put *before* the element they modify.
+    Alternatively, a modifier can be put directly before a syntactic closer or
+    separator, such as ``;`` or ``where`` or ``)``.
+
+11. Modifiers of unknown or polymorphic kind produce an error.
+
+12. Modifiers of known kind but with an unknown meaning produce a warning,
+    controlled by ``-Wunrecognized-modifiers``. They are otherwise ignored.
+    (However, in order to know that a modifier is unrecognized, it still must be
+    parsed, renamed, and type-checked.) ``-Wunrecognized-modifiers`` is enabled
+    by default.
+
+13. With ``-XLinearTypes``:
+
+    * A modifier of type ``Multiplicity`` changes the multiplicity of the
+      following arrow, or following pattern-bound variable of a lambda, or
+      following let or where binding, or preceding record field. Multiple
+      modifiers of type ``Multiplicity`` on the same arrow are not allowed. Any
+      other use of a modifier still has no meaning.
+
+    * The ``%1`` modifier is handled as a special case. It's renamed (and so
+      typechecked) the same as ``%One`` (using the ``One`` from ``base``), even
+      if it appears somewhere that linear modifiers aren't expected. If a user
+      does want the modifier ``1 :: Nat`` (with ``-XDataKinds), they can write
+      it as ``%01`` or ``%(1 :: Nat)``.
+
+    * The linear arrow ``a ⊸ b`` has the same meaning as ``a %1 -> b``. Other
+      modifiers are accepted: ``a %Matchable ⊸ b`` has the same meaning as ``a
+      %Matchable %1 -> b``.
```

Comment:
What if I had `a %Many %One -> b`? Is this a linear or non-linear arrow?

Reply:
With `-XLinearTypes`, that's an error. (Specified a few lines above: "Multiple modifiers of type `Multiplicity` on the same arrow are not allowed.") Without, both modifiers are ignored with a warning.